File: sor_sendrecv.c
/*
 * sor.c:
 * Successive over relaxation
 * MPI version implementing a red-black SOR, based on earlier Orca source.
 *
 * Thilo Kielmann, 12/08/1998
 *
 */

#include "stdio.h"
#include "stdlib.h"
#include "math.h"
#include "mpi.h"

int even (int i){
  return ( ( ( i / 2 ) * 2 ) == i ) ? 1 : 0;
}

double abs_d (double d){
  return (d < 0.0) ? -d : d;
}

double stencil (double**G, int row, int col){
  return ( G[row-1][col] + G[row+1][col] + G[row][col-1] + G[row][col+1] )
         / 4.0;
}

/* compute lower and upper bound of the grid stripe */
static void
get_bounds(int *lb, int *ub, int n, int size, int rank)
{
    int nlarge, nsmall;
    int size_large, size_small;

    nlarge = n % size;
    nsmall = size - nlarge;

    size_small = n / size;
    size_large = size_small + 1;

    if (rank < nlarge) {            /* I'll  have a large chunk */
	*lb = rank * size_large;
	*ub = *lb + size_large;
    } else {
	*lb = nlarge * size_large + (rank - nlarge) * size_small;
	*ub = *lb + size_small;
    }
}

int main (int argc, char *argv[]){

  int N;                            /* problem size */
  int ncol,nrow;                    /* number of rows and columns */
  double TOLERANCE = 0.0002;        /* termination criterion */
  double Gnew,r,omega,              /* some float values */
         stopdiff,maxdiff,diff;     /* differences btw grid points in iters */
  double **G;                       /* the grid */
  int i,j,phase,iteration;          /* counters */
  double t_start,t_end;             /* time values */
  int size,rank,pred,succ;          /* process ranks */
  int lb,ub;                        /* lower and upper bound of grid stripe */
  MPI_Status status;                /* dummy for MPI_Recv */

  MPI_Init(&argc,&argv);
  MPI_Comm_size(MPI_COMM_WORLD,&size);
  MPI_Comm_rank(MPI_COMM_WORLD,&rank);

  /* ranks of predecessor and successor for row exchanges */
  pred = (rank == 0) ? MPI_PROC_NULL : rank - 1;
  succ = (rank == size-1) ? MPI_PROC_NULL : rank + 1;

  /* set up problem size */
  N = 0;
  if ( argc > 1 )
    N = atoi(argv[1]);
  if (N == 0)
    N = 1000;

  if ( N < size ) N = size; /* give each process at least one row */

  if (rank == 0 ) {
      printf("Running SOR on %d nodes with %d rows\n", size, N);
  }

  N += 2; /* add the two border lines */
          /* finally N*N (from argv) array points will be computed */

  /* set up a quadratic grid */
  ncol = nrow = N;
  r        = 0.5 * ( cos( M_PI / ncol ) + cos( M_PI / nrow ) );
  omega    = 2.0 / ( 1 + sqrt( 1 - r * r ) );
  stopdiff = TOLERANCE / ( 2.0 - omega );
  omega   *= 0.8;                   /* magic factor */

  /* get my stripe bounds and malloc the grid accordingly */
  get_bounds(&lb, &ub, N-1, size, rank);
  if (lb == 0) lb = 1; /* row 0 is static */

  G = (double**)malloc(nrow*sizeof(double*));
  if ( G == 0 ){ printf("malloc failed\n"); exit(42); }
  for (i = lb-1; i<=ub; i++){ /* malloc the own range plus one more line */
                              /* of overlap on each border */
    G[i] = (double*)malloc(ncol*sizeof(double));
    if ( G[i] == 0 ){ printf("malloc failed\n"); exit(42); }
  }

  /* initialize the grid */
  for (i = lb-1; i <= ub; i++){
    for (j = 0; j < ncol; j++){
      if (i == 0) G[i][j] = 4.56;
      else if (i == nrow-1) G[i][j] = 9.85;
      else if (j == 0) G[i][j] = 7.32;
      else if (j == ncol-1) G[i][j] = 6.88;
      else G[i][j] = 0.0;
    }
  }

  /* now do the "real" computation */
  t_start = MPI_Wtime();
  iteration = 0;
  do {
    MPI_Send(G[lb]  ,ncol,MPI_DOUBLE,pred,42,MPI_COMM_WORLD);
    MPI_Send(G[ub-1],ncol,MPI_DOUBLE,succ,42,MPI_COMM_WORLD);
    MPI_Recv(G[lb-1],ncol,MPI_DOUBLE,pred,42,MPI_COMM_WORLD,&status);
    MPI_Recv(G[ub]  ,ncol,MPI_DOUBLE,succ,42,MPI_COMM_WORLD,&status);
    maxdiff = 0.0;
    for ( phase = 0; phase < 2 ; phase++){
      for ( i = lb ; i < ub ; i++ ){
        for ( j = 1 + (even(i) ^ phase); j < ncol-1 ; j += 2 ){
          Gnew = stencil(G,i,j);
          diff = abs_d(Gnew - G[i][j]);
          if ( diff > maxdiff )
            maxdiff = diff;
          G[i][j] = G[i][j] + omega * (Gnew-G[i][j]);
        }
      }
    }
    diff = maxdiff;
    MPI_Allreduce(&diff,&maxdiff,1,MPI_DOUBLE,MPI_MAX,MPI_COMM_WORLD);
    iteration++;
  } while (maxdiff > stopdiff);
  t_end = MPI_Wtime();

  if ( rank == 0 ){
    printf("SOR %d x %d took %f s\n",nrow-2,ncol-2,t_end-t_start);
    printf("using %d iterations, diff is %f (allowed diff %f)\n",
           iteration,maxdiff,stopdiff);
  }

  MPI_Finalize();
  return 0;
}

File: src/sor_sendrecv.rs
use std::env;
use std::f64::consts::PI;
use mpi::collective::SystemOperation;
use mpi::Rank;
use mpi::topology::SimpleCommunicator;
use mpi::traits::*;
use crate::test_utils::powers_of_two;

fn even_1_odd_0(num: usize) -> usize {
    match num % 2 {
        0 => 1,
        _ => 0
    }
}

// Suppose n = 1000, size = 4, rank = 2
fn get_bounds(n: usize, size: usize, rank: usize) -> (usize, usize) {
    let mut nlarge = n % size; // 1000 % 4 = 0
    let mut nsmall = size - nlarge; // 4 - 0 = 4

    let mut size_small = n / size; // 1000 / 4 = 25
    let  size_large = size_small + 1; // 25 + 1 = 26

    let mut lower_bound;
    let mut upper_bound;

    if rank < nlarge { // 2 < 0 ?
        lower_bound = rank * size_large;
        upper_bound = lower_bound + size_large;
    } else {
        // 0 * 26 + (2 - 0) * 4 = 8
        lower_bound = nlarge * size_large + (rank - nlarge) * size_small;
        // 8 + 25 = 33
        upper_bound = lower_bound + size_small;
    }
    (lower_bound, upper_bound)
}

pub fn runner(problem_size: usize) {
    let universe = mpi::initialize().unwrap();
    let world = universe.world();
    let rank = world.rank();

    for n in powers_of_two(problem_size as u32) {
        sor(n as usize, rank, &world);
    }
}

pub fn sor(problem_size: usize, rank: Rank, world: &SimpleCommunicator) {
    let size = world.size();

    let pred_rank = if rank == 0 { rank } else { rank - 1 };
    let succ_rank = if rank == size - 1 { rank } else { rank + 1 };

    // println!("I'm rank {}, my pred rank: {}, succ rank: {}", rank, pred_rank, succ_rank);
    // initialize the basic variables
    let mut N = problem_size;
    if N == 0 {
        N = 1000;
    }
    // Give each process at least one row
    if N < size as usize {
        N = size as usize;
    }

    if rank == 0 {
        println!("Running SOR on nodes: {}, rows: {}", size, N);
    }
    N += 2;

    let n_col = N;
    let n_row = N;
    let r = 0.5 * ((PI / n_col as f64).cos() + (PI / n_row as f64).cos());
    let mut  omega =  2.0 / (1.0 + (1.0 - r * r).sqrt());
    let stop_diff = 0.0002 / (2.0 - omega);
    // println!("Stop diff: {}", stop_diff);
    let mut max_diff;
    let mut diff;
    omega *= 0.8;

    // get my stripe bounds and malloc the grid accordingly
    let (mut global_lb, mut global_ub) = get_bounds(N - 1, size as usize, rank as usize);
    // row 0 is static
    if global_lb == 0 {
        global_lb = 1;
    }
    // Initialize the matrix at local rank, full size, 0 filled
    let local_ub = global_ub - (global_lb - 1);
    let mut matrix = vec![vec![0.0; n_col]; local_ub+1];
    // Initialize the boundary value
    for i in 0..=local_ub {
        for j in 0..n_col {
            if i == 0 && global_lb - 1 == 0 {
                matrix[i][j] = 4.56;
            } else if i == local_ub && global_ub == n_row - 1 {
                matrix[i][j] = 9.85;
            } else if j == 0 {
                matrix[i][j] = 7.32;
            } else if j == n_col - 1 {
                matrix[i][j] = 6.88;
            } else {
                matrix[i][j] = 0.0;
            }
        }
    }

    let t_start = mpi::time();
    // Now do the real computation
    let mut iteration = 0;
    loop {
        if pred_rank != rank {
            world.process_at_rank(pred_rank).send_with_tag(&matrix[1], 42);
            world.process_at_rank(pred_rank).receive_into_with_tag(&mut matrix[0], 42);
        }
        if succ_rank != rank {
            world.process_at_rank(succ_rank).send_with_tag(&matrix[local_ub -1], 42);
            world.process_at_rank(succ_rank).receive_into_with_tag(&mut matrix[local_ub], 42);
        }

        max_diff = 0.0;
        for phase in 0..2 {
            let mut global_row_num = global_lb;
            for i in 1..local_ub {
                let start_col = 1 + (even_1_odd_0(global_row_num) ^ phase);
                for j in (start_col..n_col-1).step_by(2) {
                    // The stencil operation
                    let up = &matrix[i - 1][j];
                    let down = &matrix[i + 1][j];
                    let left = &matrix[i][j - 1];
                    let right = &matrix[i][j + 1];

                    let stencil_val = (up + down + left + right) / 4.0;
                    diff = (stencil_val as f64 - matrix[i][j] as f64).abs();

                    if diff > max_diff {
                        max_diff = diff;
                    }
                    matrix[i][j] = matrix[i][j] + omega * (stencil_val - matrix[i][j]);
                }
                global_row_num += 1;
            }
        }

        diff = max_diff;
        world.all_reduce_into(&diff, &mut max_diff, SystemOperation::max());
        iteration += 1;

        if max_diff <= stop_diff {
            break;
        }
    }
    let t_end = mpi::time();

    if rank == 0 {
        println!("SOR size: {} x {}, time: {} ms", n_row-2, n_col-2, (t_end-t_start) * 1000.0);
        println!("using {} iterations, diff is {} (allowed diff {})", iteration,max_diff,stop_diff)
    }
}

fn print_matrix(matrix: &Vec<Vec<f64>>, n_row: usize, n_col: usize, rank: Rank) {
    println!("Rank {}'s matrix -----------", rank);
    for i in 0..n_row {
        print!("r[{}]: ", i);
        for j in 0..n_col {
            print!(" {} ", matrix[i][j]);
        }
        println!();
    }
}
// try from 10x10, check matrix if identical
// iterations should be same
// only use boundary rows
// check out if optimization options